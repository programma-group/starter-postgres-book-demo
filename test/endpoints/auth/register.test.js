const request = require('supertest');
const app = require('../../../src/app');
const User = require('../../../src/models/user');
const { getCorrectUserData, getOmittedData } = require('../../utils/user');

const makeRequest = (body = {}) => (
  request(app)
    .post('/auth/register')
    .send(body)
);

const requiredFields = [
  'username',
  'password',
  'passwordConfirm',
  'email',
  'firstName',
  'lastName',
];

describe('POST /auth/register', () => {
  it('should register a user', async () => {
    const data = getCorrectUserData();
    await makeRequest(data)
      .expect((res) => {
        expect(res.body).toMatchObject(expect.objectContaining({
          ok: true,
          data: {
            token: expect.any(String),
          },
        }));
      })
      .expect(200);
    const c = await User.query().where('email', data.email).count();
    expect(c[0].count).toBe('1');
    const userObj = await User.query().where('email', data.email).first();
    expect(userObj).toMatchObject(expect.objectContaining({
      email: data.email,
      firstName: data.firstName,
      lastName: data.lastName,
      username: data.username,
    }));
  });
  requiredFields.forEach((field) => {
    const data = getOmittedData(field);
    it(`should return error if ${field} is not present on the payload`, async () => {
      const initialCount = await User.query().count();
      await makeRequest(data)
        .expect((res) => {
          expect(res.body).toMatchObject(expect.objectContaining({
            ok: false,
            type: 'BodyValidationError',
            message: 'There are validation errors',
          }));
          expect(res.body.errors).toMatchSnapshot(`${field} errors`);
        })
        .expect(400);
      const finalCount = await User.query().count();
      expect(initialCount[0].count).toBe(finalCount[0].count);
    });
  });
  it('should not allow email and username twice', async () => {
    const data = getCorrectUserData();
    await makeRequest(data).expect(200);
    await makeRequest(data)
      .expect((res) => {
        expect(res.body).toMatchObject(expect.objectContaining({
          ok: false,
          type: 'ModelValidation',
        }));
        expect(Object.keys(res.body.data)).toHaveLength(2);
        const { email, username } = res.body.data;
        expect(email).toMatchSnapshot('email');
        expect(username).toMatchSnapshot('username');
      })
      .expect(400);
  });
});
