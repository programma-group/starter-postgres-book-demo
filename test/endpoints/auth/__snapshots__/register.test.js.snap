// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /auth/register should not allow email and username twice: email 1`] = `
Array [
  Object {
    "keyword": "unique",
    "message": "email already in use.",
  },
]
`;

exports[`POST /auth/register should not allow email and username twice: username 1`] = `
Array [
  Object {
    "keyword": "unique",
    "message": "username already in use.",
  },
]
`;

exports[`POST /auth/register should return error if email is not present on the payload: email errors 1`] = `
Array [
  Object {
    "location": "body",
    "msg": "That Email is not valid!",
    "param": "email",
  },
]
`;

exports[`POST /auth/register should return error if firstName is not present on the payload: firstName errors 1`] = `
Array [
  Object {
    "location": "body",
    "msg": "First name cannot be blank!",
    "param": "firstName",
  },
]
`;

exports[`POST /auth/register should return error if lastName is not present on the payload: lastName errors 1`] = `
Array [
  Object {
    "location": "body",
    "msg": "Last name cannot be blank!",
    "param": "lastName",
  },
]
`;

exports[`POST /auth/register should return error if password is not present on the payload: password errors 1`] = `
Array [
  Object {
    "location": "body",
    "msg": "Password cannot be blank!",
    "param": "password",
  },
  Object {
    "location": "body",
    "msg": "Passwords don't match",
    "param": "passwordConfirm",
    "value": "123456",
  },
]
`;

exports[`POST /auth/register should return error if passwordConfirm is not present on the payload: passwordConfirm errors 1`] = `
Array [
  Object {
    "location": "body",
    "msg": "Confirmed password cannot be blank!",
    "param": "passwordConfirm",
  },
  Object {
    "location": "body",
    "msg": "Passwords don't match",
    "param": "passwordConfirm",
  },
]
`;

exports[`POST /auth/register should return error if username is not present on the payload: username errors 1`] = `
Array [
  Object {
    "location": "body",
    "msg": "Username cannot be empty",
    "param": "username",
  },
]
`;
